src/loki/tile_action.cc

```diff
@@ -870,20 +870,78 @@ void build_layers(const std::shared_ptr<GraphReader>& reader,
 namespace valhalla {
 namespace loki {
 
+namespace detail {
+std::filesystem::path
+mvt_local_path(const uint32_t z, const uint32_t x, const uint32_t y, const std::string& root) {
+  static std::string kMvtExt = ".mvt";
+  // number of cols & rows
+  size_t dim = 1ull << z;
+
+  // determine zero padded width for the full path
+  size_t max_index = (dim * dim) - 1;
+  size_t path_width = static_cast<size_t>(std::log10(max_index)) + 1;
+  const size_t remainder = path_width % 3;
+  if (remainder) {
+    path_width += 3 - remainder;
+  }
+  assert(path_width % 3 == 0);
+
+  // convert index to zero-padded decimal string
+  size_t tile_index = static_cast<size_t>(y) * dim + static_cast<size_t>(x);
+  std::ostringstream oss;
+  oss << std::setw(path_width) << std::setfill('0') << tile_index;
+  std::string path_no_sep = oss.str();
+
+  // split into groups of 3 digits
+  std::vector<std::string> groups;
+  size_t i = 0;
+  while (i < path_no_sep.size()) {
+    groups.push_back(path_no_sep.substr(i, 3));
+    i += 3;
+  }
+
+  std::filesystem::path tile_path = root;
+  tile_path /= std::to_string(z);
+
+  // append all groups but the last one, which is the filename
+  for (size_t i = 0; i < groups.size() - 1; ++i) {
+    tile_path /= groups[i];
+  }
+  tile_path /= groups.back() + kMvtExt;
+
+  return tile_path;
+}
+} // namespace detail
+
 std::string loki_worker_t::render_tile(Api& request) {
   const auto& options = request.options();
 
   vtzero::tile_builder tile;
   const auto z = options.tile_xyz().z();
-  if (z < min_zoom_) {
+  if (z < min_zoom_road_class_.front()) {
     return tile.serialize();
+  } else if (z > min_zoom_road_class_.back()) {
+    // throwing allows clients (mapblibre at least) to overzoom
+    throw valhalla_exception_t{175, std::to_string(min_zoom_road_class_.back())};
   }
 
   // time this whole method and save that statistic
   auto _ = measure_scope_time(request);
 
+  // do we have it cached?
+  const auto x = options.tile_xyz().x();
+  const auto y = options.tile_xyz().y();
+  const auto tile_path = detail::mvt_local_path(z, x, y, mvt_cache_dir_);
+  bool is_cached = std::filesystem::exists(tile_path);
+  bool cache_allowed = (z >= mvt_cache_min_zoom_) && !mvt_cache_dir_.empty();
+
+  if (is_cached && cache_allowed) {
```

Comment:
we should probably reorganize this just a tiny bit so that cache allowed is first in the if and only if its allow we actually call filesystem::exists, since that is a sys call and could be expensive if its not a filesystem cache hit. microoptimization i know but not difficult to quickly change

Reply:
jep, done